### PR TITLE
Delta key file download and submission (EXPOSUREAPP-12213)

### PR DESCRIPTION
**Pull request by chiljamgossow**
- download only files that are missing and have not been checked
- mark key files as checked after successful submission
- submit only delta files

- [X] unit test delta download
- [X] unit test update checked flag

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/BaseKeyPackageSyncTool.kt [hunk 1]
```diff
@@ -87,19 +87,20 @@ open class BaseKeyPackageSyncTool(
...
             val exists = key.path.exists()
-            if (complete && !exists) {
+            val checked = key.info.checkedForExposures
+            if (complete && !exists && !checked) {
                 Timber.tag(tag).v("Incomplete download, will overwrite: %s", key)
             }
-            // We overwrite not completed ones
-            complete && exists
+            // We overwrite incomplete ones that have not been checked
+            complete && (exists || checked)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/BaseKeyPackageSyncTool.kt:103]: These 4 lines are confusing, in `if` checks for `!` existing `!` checked keys and logs call them incomplete:
```kotlin
if (complete && !exists && !checked) {
   Timber.tag(tag).v("Incomplete download, will overwrite: %s", key)
}
```
BUT in here, we return complete and (existing or checked)  and the comment says incomplete too:
```kotlin 
// We overwrite incomplete ones that have not been checked
complete && (exists || checked)
```
Could you please elaborate more ? or correct the log / comment statements?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt [hunk 2]
```diff
@@ -92,25 +92,25 @@ class DayPackageSyncTool @Inject constructor(
...
-        val staleDays = cachedDays.findStaleData(listOf(availableDays))
-
-        if (staleDays.isNotEmpty()) {
-            Timber.tag(TAG).d("Deleting stale days (loation=%s): %s", location, staleDays)
-            keyCache.delete(staleDays.map { it.info })
+        // remove files that are no longer available on the server
+        val staleKeys = cachedKeys.findStaleData(listOf(availableDays))
+        if (staleKeys.isNotEmpty()) {
+            Timber.tag(TAG).d("Deleting stale days (loation=%s): %s", location, staleKeys)
+            keyCache.delete(staleKeys.map { it.info })
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt:109]: - is `checkedForSubmission` Keys part of this? , if so, we should `NOT` delete the `meta-data` as per specs , but we SHOULD delete the `files`
- I don't see where actually the `files` are deleted. Do we delete them anywhere?

**Review by thomasaugsten — APPROVED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt:109]: Stale Keys are those which have been removed from the server. They cannot be downloaded anymore and should also be removed from the device.
Delete will remove the meta data and the file itself.

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt:109]: ok, where do we delete checkedForSubmission files?

Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DayPackageSyncTool.kt:109]: next PR

**Review by mtwalli — APPROVED** [review]
Thanks!

**Pull request merged**